Added ODT support for spans and containers (divs).

LarsGit223 opened the pull request:
Here is a simple implementation of ODT support for spans and containers (divs).

As you possibly noticed I did write a patch for ODT support before but that was very ugly. All the ODT knowledge was included in the wrap plugin itself. Including ugly arrays to keep the style information.

Know that I wrote a little CSS import helper class for the ODT plugin the code looks much nicer now.

Of course it is far from perfect. One reason is simply the differences between XHTML and the ODT format.

But maybe you like it enough to merge it. This would be great, then I don't need to keep my own branch.

selfthinker commented:
Thanks. This would be a great enhancement, many people keep on asking for ODT support.

But it's not working for me. When I try to export the example page (the one that comes with the plugin as example.txt) I get this error when I try to open the resulting file:

> Read-Error.
> Format error discovered in the file in sub-document content.xml at 624,491(row,col).

I tried a different page with fewer and more basic wrap examples and I get the same error (different row, col obviously). The exported wiki:syntax page looks okay.
(Not sure if it's important, I use LibreOffice 4.3.5.2.)

selfthinker commented on syntax/div.php:
```diff
@@ -108,9 +109,58 @@ function render($mode, Doku_Renderer $renderer, $indata) {
     }
 
-
+    function _odtVersionToOld (Doku_Renderer &$renderer) {
```
There is a typo, it should say that the version is **too** old.

Shouldn't this method rather be part of the helper plugin? That way it will be easier to use in the span component as well.
It is generally a good practice to check for the existence of methods rather than version numbers. I don't really know the ODT plugin, so don't know if there is a valid reason for not doing that...?

LarsGit223 commented:
1. My test pages work fine but they have only text spans and highlight boxes in it. I assume that some elements which are not supported for ODT export cause the problem. Could you please provide me your example wiki source code.

2. You wrote:
> It is generally a good practice to check for the existence of methods rather than version numbers.

The reason for not doing that is simply not knowing that it is possible. I never used PHP before starting to work on DokuWiki/the ODT plugin. How do I check if a method exists in PHP?

selfthinker commented:
1. As I wrote, I use "the one that comes with the plugin as example.txt" which is [here](https://github.com/selfthinker/dokuwiki_plugin_wrap/blob/master/example.txt).

2. You can use [`method_exists()`](http://php.net/method_exists) on a class method or [`function_exists()`](http://php.net/function_exists) otherwise.

LarsGit223 commented:
2. Thanks a lot. I will build that in later.

1. As I assumed there was a problem with the not supported wrap classes. So far only spans and boxes are supported. I changed and commited the code. It works so far that the resulting ODT document can be opened.

But your example page shows an error in the ODT plugin. The heading is opened in a paragraph and so you do not see the heading text in the boxes. Change the text to normal text, no heading, then you will see it.

Your example page is a good test for the ODT plugin. Unfortunately it shows that there is still much work to do.

selfthinker commented:
This works for me now, but as you wrote, there are still some issues.
I don't mind if some classes are unsupported as long as at least the content is still shown (which it is). But the ones that are supported shouldn't look broken.

[This list](https://www.dokuwiki.org/plugin:wrap#patch_for_odt_support) is not up-to-date anymore, is it? That was for a previous version? It would be great to have a similar list for the current implementation.

LarsGit223 commented:
Arghhh...of course, you are right.
I forgot that some values are not taken out from the CSS files but from the style="..." string.

I just overhauled the code. I think I fixed the following things which did not work since the re-design:
- alignment: left, center, right
- size in %, absolute values are converted to 100%
- onlyprint (only for boxes, unfortunately LibreOffice produces a white space in the document on the screen, will re-check with OpenOffice). But it works so far that the box is only printed, see print preview.
- added missing boxes for 'hi' and 'spoiler' (spoiler for boxes is missing the red border, do'nt no yet why), spoiler for spans is fine including red border
- nested boxes (as frames) are not supported right now (and not possible in ODT). Before the nested box content was missing. Now it is at least display - but as part of the outer box. See the section ''Round tip box underneath, after a clear." in your example page. That disappeared before - now fixed.

I hope and think we have the same functionality again as before the re-design. I will now overwork and update the list on the Wikipage. Please let me know if you still see any errors/broken things.

Important: you have to check out the not yet released master branch of the ODT plugin!

Cheers, Lars

selfthinker commented:
I see quite a different outcome than you describe. Not sure if it's because I use LibreOffice? (I updated my version now to 4.3.7.2.).

* boxes are still quite buggy
* alignment is not supported in my version
* onlyprint does not work
* and more (see below)

I have listed everything I'd like to see implemented (with MoSCoW priorities) and the bugs I found so far in a Google spreadsheet:
https://docs.google.com/spreadsheets/d/1F4OE4Kc9TzV3JwXxJc62TgxSbkPmOuCNmfgyzWp3Xco/edit?usp=sharing (Let me know if you'd like write access to that.)

As soon as every M is implemented and nothing is too buggy, I'd be happy to merge this.
It's a bit difficult if we both seem to be seeing different things. ;-) What are you testing with?

LarsGit223 commented:
OK, here is what I am working with:

- Ubuntu 14.04 LTS
- Apache2
- DokuWiki Hrun (d I think)
- wrap plugin: https://github.com/LarsGit223/dokuwiki_plugin_wrap from 2015-05-09
- odt plugin: https://github.com/LarsGit223/dokuwiki-plugin-odt/commits/master from 2015-05-09 (but 2015-05-07 should also be fine. There were no wrap specific changes in 2015-05-09)
- LibreOffice version 4.2.8.2

LarsGit223 commented:
WOOOW! Just updated to LibreOffice 4.4.3.2. Now I see what you mean :-)

Compared to my previous results with LibreOffice 4.2.8.2 I see the following differences now:
(Legend:
(-) = negative difference, worked before
(*) = neutral, no change,
(+) = positive change, did not work before)

(-) boxes clash into each other (they did'nt before)

(-) boxes don't have the correct size (worked before)

(-) alignment worked before. I noticed that e.g. if I change the alignment of the "important" box to right, than the boxes do not clash anymore. Strange. (I mean alignment just left right center worked before - also the boxes never floated cause floating is not implemented)

(*) Multi columns still seem to work (since 2015-05-07, not implemented before)

(-) Onlyprint only works for BOXES not for spans. But seems to have changed right now. Not showns on the screen but also not shown properly in print preview

(*) spans still seem to look good so far (except the little issues regarding the icons)

(+) spans nested in boxes now have the right background-color. In my previous test the background color behind the text span "emphasized and highlighted with a nested notice inside" was only grey, just like the surrounding box. Not is is yellow and blue as it should be. Before I did not even have the option in the LibreOffice GUI to change the background color of the text in the frame. So this is obviously something that they fixed.

Honestly, I would have never expected such a big difference between the LibreOffice versions.

I also would like to comment things on the list, so please give me write access.

Thanks for investing your time in this. I did not change the wrap patch for a long time before cause I had the impression that there is no acceptance/interest in merging it.

Sorry for the boxes. They have never ever looked so bad before with LibreOffice 4.2. I probably wouldn't even have put my patch on the wrap page if I knew that boxes look so bad :-)

selfthinker commented:
> Honestly, I would have never expected such a big difference between the LibreOffice versions.
Yes, that's quite surprising. I thought differences between LibreOffice and OpenOffice or Linux and Windows could potentially be greater. But these are differences between only **minor** versions... :(

> I also would like to comment things on the list, so please give me write access.
Done.
Just add new columns like H and I for each different version.

> Thanks for investing your time in this. I did not change the wrap patch for a long time before cause I had the impression that there is no acceptance/interest in merging it.
There is a lot of interest, but I generally neglected all things DokuWiki for the last few months and only now started to catch up. Sorry for the delay. I definitely appreciate your work!

selfthinker commented:
> Just add new columns like H and I for each different version.

As I now deleted some columns, those are now F and G. So, basically the two columns under "current support".

LarsGit223 commented:
Ähhhh...stupid question:
Never used google office/tables before. What account credentials (login/password) do I have or better say to whom did you grant write access?

selfthinker commented:
Hmm, weird I replied via email 2 hours ago (I don't do that often) and my reply is not showing here...
This is what I wrote:

You should have got an email with instructions from the system. If you haven't, check your spam folder.

LarsGit223 commented:
Thanks, just saw it now. I did not have my e-mail open.

I fixed some things, will also update the list:
- size of boxes (odt plugin)
- multi columns for 2 columns fixed (odt plugin)
- support for hide (display-attribute) in text spans and paragraphs
- added alignment for paragraphs (wrap plugin)

selfthinker commented on syntax/div.php:
```diff
@@ -98,6 +105,8 @@ function render($mode, Doku_Renderer $renderer, $indata) {
                     $wrap =& plugin_load('helper', 'wrap');
                     $attr = $wrap->buildAttributes($data, 'plugin_wrap');
 
+                    $renderer->doc .= 'ATTR:'.$attr.'!';
```
You left some debugging code.

selfthinker commented on syntax/div.php:
```diff
@@ -40,7 +46,7 @@ function postConnect() {
      * Handle the match
      */
-    function handle($match, $state, $pos, Doku_Handler $handler){
+    function handle($match, $state, $pos, Doku_Handler &$handler){
```
I had just removed the `&`, it is not necessary in PHP5 anymore and was throwing a "strict standards" error (see #72).

selfthinker commented on syntax/div.php:
```diff
@@ -78,7 +84,8 @@ function handle($match, $state, $pos, Doku_Handler $handler){
      * Create output
      */
-    function render($mode, Doku_Renderer $renderer, $indata) {
+    function render($mode, Doku_Renderer &$renderer, $indata) {
```
The same here, the `&` should be removed.

selfthinker commented on syntax/div.php:
```diff
@@ -98,6 +105,8 @@ function render($mode, Doku_Renderer $renderer, $indata) {
                     $wrap =& plugin_load('helper', 'wrap');
```
The `&` should be removed here as well.

selfthinker replied:
Oh, I just realised that wasn't you. Just merge my master into your branch and that will be removed (because I removed it in #86).

selfthinker commented on syntax/div.php:
```diff
@@ -108,9 +117,377 @@ function render($mode, Doku_Renderer $renderer, $indata) {
+                case DOKU_LEXER_ENTER:
+                    // Does CSS-Import exist?
+                    if ( $this->odt_ignore === false && self::$import == NULL ) {
+                        self::$import =& plugin_load('helper', 'odt_cssimport');
```
And here (`&`)...

selfthinker commented on syntax/div.php:
```diff
@@ -108,9 +117,377 @@ function render($mode, Doku_Renderer $renderer, $indata) {
+                    }
+
+                    // Get attributes. Use the same mode as for XHTML.
+                    $wrap =& plugin_load('helper', 'wrap');
```
And here (`&`)...

selfthinker commented on syntax/span.php:
```diff
@@ -40,7 +40,7 @@ function postConnect() {
      * Handle the match
      */
-    function handle($match, $state, $pos, Doku_Handler $handler){
+    function handle($match, $state, $pos, Doku_Handler &$handler){
```
And here (`&`)...

selfthinker commented on syntax/span.php:
```diff
@@ -60,7 +60,8 @@ function handle($match, $state, $pos, Doku_Handler $handler){
      * Create output
      */
-    function render($mode, Doku_Renderer $renderer, $indata) {
+    function render($mode, Doku_Renderer &$renderer, $indata) {
```
And here (`&`)...

selfthinker commented on syntax/span.php:
```diff
@@ -80,9 +81,163 @@ function render($mode, Doku_Renderer $renderer, $indata) {
+            switch ($state) {
+                case DOKU_LEXER_ENTER:
+                    // Get attributes.
+                    $wrap =& plugin_load('helper', 'wrap');
```
And here (`&`)...

selfthinker commented on syntax/span.php:
```diff
@@ -80,9 +81,163 @@ function render($mode, Doku_Renderer $renderer, $indata) {
+
+                    // Import Wrap-CSS.
+                    if ( self::$import == NULL ) {
+                        self::$import =& plugin_load('helper', 'odt_cssimport');
```
And here (`&`)...

selfthinker commented:
Wow, what a difference! This current version is so much better. :)

I think it's nearly ready, the only things I would like to have fixed are:

* Formatting in boxes should work. I saw your comment about potential bugs in LibreOffice. If it's not possible or otherwise too difficult, I would be fine with merging this without a fix.
* The support for the emulated headings would probably be tied to that? I just found that the "big emulated heading" at least shows some underlining. But the bold is missing from both.

I also replied to some of your comments in the spreadsheet and removed some other comments about features which were clearly implemented.

LarsGit223 commented:
Oooops, sorry for leaving the debug code.

Yes, I saw the differences with the '&'s too. Anyway, before fixing together a final version I would check out the latest version anyway and put my changes in there.

I assume that the emulated heading is caused by the general formatting problem in boxes. Also I have to admit that I didn't find the code that is actually handling the emulated headlines. Where is that 'magic' happening?

I noticed that the styles for headings are paragraph styles. So as a workaround I could insert normal paragraphs with the headings style. If that works it would be the same look but semantically not a real headline.

You maybe noticed that in a text box you can not select any predefined format styles. But you can do direct style changes. That is another reason why I think it is a bug in LibreOffice. Also, in 4.2 that I used before the background-colors didn't work in the wrp example page with that highlighted text and the notice incuded in the grey box. Looks like they still have work to do there.

One question:
The noprint would be something that could be implemented easily. Is that something that I can see from the CSS properties? I wonder that 'wrap_onlyprint' and 'wrap_noprint' have the same value for the display property. So I guess I have to query the class name directly and adjust the ODT display property accordingly.

selfthinker commented:
> I assume that the emulated heading is caused by the general formatting problem in boxes. Also I have to admit that I didn't find the code that is actually handling the emulated headlines. Where is that 'magic' happening?

It's all just CSS, see https://github.com/selfthinker/dokuwiki_plugin_wrap/blob/master/all.css#L16-30 (below those lines are also special cases of the same).

> I noticed that the styles for headings are paragraph styles. So as a workaround I could insert normal paragraphs with the headings style. If that works it would be the same look but semantically not a real headline.

The question is which heading styles you would use? Wouldn't that be different per template? I'd be happy enough with just a bold, seeing that hardly any other formatting seems to be working. E.g. the first box under "Boxes and Notes" doesn't show the list formatting or not even any of the code/monospaced text (or bold if I add any). And there doesn't seem to be any spacing either.

> One question:
The noprint would be something that could be implemented easily. Is that something that I can see from the CSS properties? I wonder that 'wrap_onlyprint' and 'wrap_noprint' have the same value for the display property. So I guess I have to query the class name directly and adjust the ODT display property accordingly.

Sorry, I don't understand the question...
Do you mean why both are set to `display: none;`? One is set in `screen` styles, and the other in `print` styles.

LarsGit223 commented:
Yes, that was exactly what I meant. I didn't see the '@media' annotation for 'screen' but found it in the resulting, temporary CSS code of the wiki page. But I didn't wanted to import the whole thing. I now simple queried the class and set the ODT 'display' property accordingly to 'always', 'screen' or 'printer'.

The workaround I tried for the box formatting problem didn't work. Also the styles.xml/template is not parsed so I can not simply copy the style e.g. for a heading to an auotmatic style.

So I decided to leave this as it is. If I find any options for improvement in the future then it will be done.

I will now close this pull-request and create a new one. I did a fresh fork/checkout from the wrap plugin and put in my changes there. So you will get a one-shot change without all the "plumbering" change messages.
